proposals/low-level-struct-improvements.md

```diff
@@ -50,248 +45,226 @@ readonly ref struct Span<T>
+{
+    // The implementation of this method is irrelevant when considering the lifetime of the 
+    // returned Span<T>. The rules disallow capture of `parameter` hence the return of 
+    // CreateSpan<T> is always safe-to-escape to the calling method
 }
 ```
 
-At the same time it is legal to have methods today which take a `ref` parameter and return a `Span<T>`.  These methods bear a lot of similarity to the newly added `Span<T>` constructor: take a `ref`, return a `Span<T>`. However the lifetime of the return value of these methods is never restricted by the inputs.  The existing span safety rules consider such values as effectively always *safe-to-escape* outside the enclosing method.
+By the existing span safety rules the return of an invocation of this method always has a //*safe-to-escape* value of *calling method*. That is because in the span safety document `ref` fields do not exist and hence there is no way for the return to capture `parameter` hence there is no constraint on the return.  The implementation of the method is completely irrelevant here, which is why it's omitted from the sample, because there is simply no way for the `ref` to be captured. This means no matter how this method is called the return is *safe-to-escape* to the *calling method*.
 
-```cs
-class ExistingScenarios
+```c#
+Span<T> CompatExample(ref int p)
 {
-    Span<T> CreateSpan<T>(ref T p)
-    {
-        // The implementation of this method is irrelevant. From the point of
-        // the consumer the returned value is always safe to return.
-        ... 
-    }
-
-    Span<T> Examples<T>(ref T p, T[] array)
-    {
-        // Legal today
-        return CreateSpan(ref p); 
+    // Okay
+    int local = 42;
+    return CreateSpan<int>(ref local);
 
-        // Legal today, must remain legal
-        T local = default;
-        return CreateSpan(ref local);
+    // Okay
+    Span<T> span = stackalloc int[1];
+    return CreateSpan<int>(ref span[0]);
 
-        // Legal for any possible value that could be used as an argument
-        return CreateSpan(...);
-    }
+    // Okay
+    int[] array = new int[42];
+    return CreatSpan(ref array[0]);
 }
 ```
 
-The reason that all of the above samples are legal is because in the existing design there is no way for the return `Span<T>` to store a reference to the input state of the method call. This is because the span safety rules explicitly depend on `Span<T>` not having a constructor which takes a `ref` parameter and stores it as a field. 
+All of the above `return` statements are legal by our existing rules because the return of `CreateSpan` is always *safe-to-escape* to the *calling method*. The challenge in this proposal, which comes up many times, is they **must** remain legal in the version of the language which implements these features and / or when `Span<T>` moves to using `ref` fields. The above patterns can legally exist today and cannot become errors when moving to a new language version. 
```

Comment:
> and cannot become errors when moving to a new language version. 

Would it be ok for them to become errors if there was true memory unsafety before?

Reply:
Warning wave maybe?

There is, on the BCL side, the interesting case of `MemoryMarshal.CreateSpan()` which functionally works like the proposed constructor (and via an internal `Span(ref T ptr, int length)` constructor):
```csharp
[RefFieldEscapes]
public Span(ref T value) { }
```

The BCL will need to decide how it works moving forward, including whether it starts erroring just like the proposed constructor, or continues allowing users to do "unsafe" things (potentially with an analyzer, etc).

Reply:
> Would it be ok for them to become errors if there was true memory unsafety before?

This pattern is completely safe though. The only way this could be unsafe is if `Span<T>` had a constructor in the form of `Span(ref T value)` which is explicitly disallowed by the span safety rules. Yes `Span<T>` does have this CTOR but it's `internal` and known to be `unsafe`. 

So there is nothing wrong with these patterns. They are safe, legal and honestly expected. The issue is that we must maintain those properties going forward.

Reply:
> Warning wave maybe?

How would a warning wave help though? Consider that by adding `[RefFieldEscapes]` to a method you change the semantics from the perspective of the caller. Or consider it this way:

```
Span<int> MyMethod() {
  int local = 42;
  Span<int> span = CreateSpan(ref local); 
  return span;
}
``` 

This is 100% legal today. If we suddenly warn customers into using `[RefFieldsEscape]` on such methods then it suddenly breaks callers of that method.

Reply:
Just so i understand with that example.  You're sayign that that code coudl exist and have legit uses (for example, not captuing 'local' within the returned span).  And as such, because of hte legal reasons, we couldn't jsut always start warning/error'ing here.

*however*, what we can do is have CreateSpan announce it's intentions about what happens with the passed in 'ref'.  If it ignored it (as an example) it would just do nothing, and the above would continue to be fine.  However, if it ends up pushing that parameter into the span it returns, then it shoudl have this attribute (virally spread hopefully from `Span<T>` inside it) so that hte caller can now error *in that case*.

--

In other words, the attribute allows cooperation between the caller and callee here on legal cases that should continue working, vs bad cases that ideally should break because they're entire memory unsafe.  Is that right?

Reply:
> This is 100% legal today.

Right, and in the case where `CreateSpan` isn't updated to use some method/constructor marked `RefFieldsEscape` to create the returned span, it continues being "safe".
- If `CreateSpan` were updated to use `[RefFieldEscapes] public Span(ref T value) { }`, then I would assume the compiler errors as it expects `CreateSpan` to then be equivalently annotated, etc. This would be a breaking change for a library to make and therefore not recommended.

There are, however, scenarios like with `MemoryMarshal.CreateSpan` where its known to be "unsafe" and where the compiler (or an analyzer) could surface a warning that it functionally works like `[RefFieldEscapes] public Span(ref T value) { }` which would help catch problematic cases (particularly for users transitioning from "old unsafe logic" to "new safe logic").

You could also consider cases like `return new Span(&local, 1)` where its unsafe and legal, but its also a detectable problem and where a warning being surfaced is probably a "good idea" (maybe these are better suited to a library analyzer though)

Reply:
> Just so i understand with that example. You're sayign that that code coudl exist and have legit uses (for example, not captuing 'local' within the returned span). And as such, because of hte legal reasons, we couldn't jsut always start warning/error'ing here.

Exactly. It's perfectly okay to create such an API and give it legitamate uses. That exact pattern exists a few places in dotnet/runtime today (`SecureString` and some parts of `DateTime` parsing I was able to find). 

> In other words, the attribute allows cooperation between the caller and callee here on legal cases that should continue working, vs bad cases that ideally should break because they're entire memory unsafe. Is that right?

yep.

Reply:
> There are, however, scenarios like with MemoryMarshal.CreateSpan where its known to be "unsafe" and where the compiler (or an analyzer) could surface a warning ... 

I'd actually vote in `net6.0` that we obsolete this API. Basically say "use `new Span<T>`" instead please. That would force the caller to use `unsafe` in the cases where actual `unsafe` behavior was still needed. 

I'm not sure how feasible that is though.

Reply:
> yep.

Gracias!  you probably did write all that out in the PR.  but it helped to get that clarity here :)